### ISSUE TITLE
Seeing "did not renew the lock" too often [NOTE: ensure 0.2.28 agent after recent sprint rollout]

### ISSUE THREAD

**Issue by joaomoreno** [Microsoft/vso-agent#74]
Every couple of days I see this error. Happens on both OS X and Linux.

```
The job has been abandoned because agent tb-lnx2 did not renew the lock.
Ensure agent is running, not sleeping, and has not lost communication
with the service.
```

![image](https://cloud.githubusercontent.com/assets/22350/9490302/4717564a-4be7-11e5-89fe-9ddf3f7d62b0.png)

**Comment by bryanmacfarlane** [comment]
We need more details.  This is not enough to go on.  It could be a bug or it could mean the agent lost communications (it renews a job every 30 seconds and server will abandon the build if not heard from in 5).  Send my team the actual build so we can look at the backend.  Also, when that occurs, look at the corresponding agent _work/_diag worker logs.

Also, ensure you're on 0.2.27 or greater.

**Comment by joaomoreno** [comment]
Looks like a service issue. Just happened on a Windows agent too. I am on `0.2.26` so I will upgrade promptly.

Nevertheless I found this in the OS X log:
```
##[Error] 2015-08-26T08:18:02.526Z: Error: Request timeout: /_apis/distributedtask/pools/5/jobrequests/1633?lockToken=00000000-0000-0000-0000-000000000000
```

**Comment by chcosta** [comment]
I am seeing this consistently in our xplat VSO builds since yesterday.  I just upgraded to version 0.2.28 today, but I am still seeing the "did not renew the lock" failure.  Who can I follow up with to resolve this and unblock our builds?

**Comment by chcosta** [comment]
@bryanmacfarlane , is there a workaround for this?  Are there logs I can provide to assist in diagnosing this or is this issue sufficiently known?

**Comment by bryanmacfarlane** [comment]
Looks to be a service side issue.

To confirm, email me and I'll get logs

Also are you blocked (all builds) ?  For others it was transient 1 or 2 builds linked to service degradation and timeouts.  Also send me your account url and we can track back on server diagnostics

**Comment by chcosta** [comment]
Thanks Bryan.  I've sent an initial email with the information you requested.  This is currently blocking us as it is affected all builds.

**Comment by jschiessl** [comment]
I can confirm that this is an issue across agent versions (0.2.14, 0.2.25) and across different build machines. So for me, this really sounds like a service side change. Worked great until Monday or Tuesday. Even previously fine running build definitions (stable for weeks and untouched for days) now get this error.
I will now hesitantly try to upgrade to latest version and see if this helps.

**Comment by jschiessl** [comment]
$(projI have now updated one of my agents to 0.2.28 but currently I am unable to determine if this fixes the problem because a new one arose:
As part of my build definition, I am calling a shell script. This script relies on a parameter being passed down so it can then build a path where this parameter is part of the path. In my case, the parameter value contains whitespace.
Previously, most of the time the parameter got passed down like this (as can be seen in the logs)
2015-08-27T09:22:02.228Z: [command]/bin/bash <pathToScript>/script.sh 'name parameter'

Now with 0.2.28 it seems that it gets always passed like this:
2015-08-27T09:22:02.228Z: [command]/bin/bash <pathToScript>/script.sh name parameter

The quotes are missing, so inside the script $1 is now 'name' instead of 'name parameter', which causes significant irritation when constructing paths ;-)
This problem sometimes occured also with previous versions, but then canceling the build and queuing it again always helped.

I am passing down the parameter as $(nameparameter) in the build definition, having it defined as name parameter (without quotes) in a variable called nameparameter and receiving it in the script as $1.
Am I doing something wrong there? If not, I should probably open a separate issue for this, right?

**Comment by bryanmacfarlane** [comment]
@jschiessl, try with double quotes.  yes, open another issue.  We're looking at the other service issue.  Upgrades are running now on the service and it might be related

**Comment by bryanmacfarlane** [comment]
We believe the solution to the problem is to update to 0.2.28

If you're hitting the issue we believe you're hitting, there was a security added to the service in M87 (just rolled out) and the agent was relying on it not being there (bug).  It was fixed in 0.2.28 so the call is made under the security context of the agent listening to the queue (as opposed to the job token).

Workaround: Another workaround (not advised - only if you're don't want to update).  Go to the pool admin UI and add user "Project Collection Build Service" to the Agent Pool Service Accounts group

**Comment by jschiessl** [comment]
Updating to 0.2.28 fixed the problem for me

**Comment by chcosta** [comment]
I updated an agent which was running on OSX as a service, I then restarted the service, but I'm still seeing the same behavior.  How do I check the version?

**Comment by jschiessl** [comment]
One way is to look at the agent dashboard at
https://yoursubdomain.visualstudio.com/_admin/_AgentPool
Then select the agent from the list and look at the Agent.NpmVersion property at the right

**Comment by chcosta** [comment]
Thanks, that confirmed that my agent version wasn't getting updated.  I was unable to successfully update the existing agents, but stopping the service, then creating a new agent installed the 0.2.28 version.  This appears to have fixed the issue for the OSX agent I tried it on.  I'll update the remaining agents we own and I'll respond back here once I've confirmed that this works for all of our agents (both OSX and Linux), so that we can close this issue.

**Comment by bryanmacfarlane** [comment]
OK - it would be good to know why the initial agent wasn't updated.  Read the update existing agent instructions on the readme carefully and let me know if you find out why.

**Comment by chcosta** [comment]
I've updated all of my agents and they appear to be operating normally now.  I would consider this issue fixed.

I'll look into why the update didn't update the current agent later, I didn't take the time to be diligent here yet.

Regarding the readme, I will note that these instructions for running the agent as a service, do not work for me..

   $ ./agent/svc.sh install

I see an error when trying to run 'svc.sh' this way, but if I make 'agent/' my working directory and just run './svc.sh install' it works just fine.

Thanks,
Chris

**Comment by bryanmacfarlane** [comment]
Cool.  I'll fix the readme.  The one issue in this thread about the occasional timeout looks to be a service issue (we looked in the service diagnostics and there's a handful of those).  Not related to the agent and the broad 0.2.28 failure.

I'll close this but wait a couple days so folks hitting this as we roll out 87 will find it easily

**Comment by joaomoreno** [comment]
Awesome, thanks for looking into it!

**Comment by hugobarona** [comment]
Hi there,

i have some agents running in the version 1.105.6, and i am getting the same issue. Is it confirmed the issue in this specific version too? Should i consider update the agents to the latest version?

Many thanks,
HB.

**Comment by ericsciple** [comment]
this agent is deprecated. there is a link from the main readme to the new agent repo (2.x agent).

**Comment by penupati** [comment]
Hi, We are getting  this "did not renew the lock" error since yesterday in our vsts linux agents. Agent details are below.
Agent.npmVeresion = 0.6.5
Agent.OS=Linux
Agent.Version=1.999.0
Let us know how to resolve this issue?

**Comment by bryanmacfarlane** [comment]
This agent was deprecated last year.  Use this one: 

https://github.com/Microsoft/vsts-agent

**Comment by SrinivasPithani** [comment]
Thanks Bryan, Would like to share some info & hopefully it can help some of other folks come around.
```
[Admin03@CI12 myagent]$ ./config.sh
Failed to load /ci12_home/Admin03/myagent/bin/libcoreclr.so, error: libunwind.so.8: cannot open shared object file: No such file or directory
Failed to bind to CoreCLR at '/ci12_home/Admin03/myagent/bin/libcoreclr.so'
[Admin03@AE-DT01-CI12 myagent]$ ls -l /ci12_home/Admin03/myagent/bin/libcoreclr.so
-rwxr--r--. 1 Admin03 Admin03 8425021 Jan  6 06:13 /ci12_home/Admin03/myagent/bin/libcoreclr.so

```
This was resolved after I have installed libunwind & libicu , both available from yum 

Also , looks like > git 2.x is mandatory   & I have resolved my issue after installing 2.8 from the source code . 
This linked helped me [ Package mgr isnt letting me to upgrade to git 2.x , so I need use the source code]
https://www.digitalocean.com/community/tutorials/how-to-install-git-on-centos-7

**Comment by rasberrypie** [comment]
My Agent version is 1.95.3 and i am still getting the error. What could be the issue?
